viewing_party/party.py

```diff
@@ -1,23 +1,161 @@
+from collections import defaultdict
+
 # ------------- WAVE 1 --------------------
 
+
 def create_movie(title, genre, rating):
-    pass
+    if title and genre and rating:
+        ans = {"title": title, "genre": genre, "rating": rating}
+        return ans
+    return None
+
+
+def add_to_watched(user_data, movie):
+    # TODO: Should we check if the movie is already in watched_movies?
+    watched_movies = user_data["watched"]  # watched_movies is a list of dictionaries.
+    watched_movies.append(movie)
+    return user_data
+
+
+def add_to_watchlist(user_data, movie):
+    # TODO: Should we check if the movie is already in the watchlist?
+    movie_watchlist = user_data["watchlist"] # movie_watchlist is a list of dictionaries.
+    movie_watchlist.append(movie)
+    return user_data
+
+
+def watch_movie(user_data, title):
+    movie_watchlist = user_data["watchlist"]
+    # TODO: Should we check if the movie has already been watched?
+    for index, movie in enumerate(movie_watchlist):
+        movie_title = movie["title"]
+        if title == movie_title:
+            add_to_watched(user_data, movie)
+            movie_watchlist[index], movie_watchlist[-1] = movie_watchlist[-1], movie_watchlist[index]
+            movie_watchlist.pop()  # considered time complexity, switch the element with last one and pop.
+            return user_data
+    return user_data   # not found
+
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
 
+def get_watched_avg_rating(user_data):
+    total_rating = 0.0
+    amount = 0
+    watched_movies = user_data["watched"]
+    for movie in watched_movies:
+        rating = movie["rating"]
+        amount += 1
+        total_rating += rating
+    if amount:  # avoid ZeroDivisionError
+        return total_rating / amount
+    return total_rating
+
+
+def get_most_watched_genre(user_data):
+    watched_movies = user_data["watched"]
+    genre_map = defaultdict(int)
+    for movie in watched_movies:
+        genre = movie["genre"]
+        genre_map[genre] += 1
+    most_watched_genre = None
+    max_amount = 0
+    for genre, times in genre_map.items():
+        # if there is a tie, keep the first
+        if times > max_amount:
+            max_amount = times
+            most_watched_genre = genre
+    return most_watched_genre
+
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
+# return a dictionary, key is title, value is a dictionary of movie's information
+def get_friends_watched_movies(friends):

```

Comment:
Good implementation. For this and similar functions, could you think about using sets to simplify some of this code?